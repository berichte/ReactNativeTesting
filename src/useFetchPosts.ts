import {useState, useEffect} from 'react';
import {Post, fetchPosts} from './fetchPosts';

export function useFetchPosts(userId: string) {
  const [data, setData] = useState<Post[]>();
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<unknown>();

  useEffect(() => {
    const fetchData = async () => {
      try {
        const posts = await fetchPosts(userId);
        setData(posts);
      } catch (e) {
        setError(e);
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [userId]);

  return {data, loading, error};
}
