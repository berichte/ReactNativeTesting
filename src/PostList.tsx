import React from 'react';
import {View, Text} from 'react-native';
import {useFetchPosts} from './useFetchPosts';

interface PostListProps {
  userId: string;
}

const PostList: React.FC<PostListProps> = ({userId}) => {
  const {data, loading, error} = useFetchPosts(userId);

  if (loading) {
    return <Text>Loading...</Text>;
  }

  if (error) {
    return <Text>Error: {(error as Error).message}</Text>;
  }

  return (
    <View>
      {data?.map(post => (
        <View key={post.id}>
          <Text>{post.title}</Text>
        </View>
      ))}
    </View>
  );
};

export default PostList;
