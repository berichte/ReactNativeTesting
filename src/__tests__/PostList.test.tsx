import React from 'react';
import {render, act} from '@testing-library/react-native';
import {useFetchPosts} from '../useFetchPosts';
import PostList from '../PostList';

jest.mock('../useFetchPosts');

describe('PostList', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render loading state', () => {
    (useFetchPosts as jest.Mock).mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });
    const {getByText} = render(<PostList userId="123" />);
    expect(getByText('Loading...')).toBeTruthy();
  });

  it('should render error state', () => {
    const errorMessage = 'Network error';
    (useFetchPosts as jest.Mock).mockReturnValue({
      data: null,
      loading: false,
      error: new Error(errorMessage),
    });
    const {getByText, toJSON} = render(<PostList userId="456" />);

    expect(getByText(`Error: ${errorMessage}`)).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render post list', async () => {
    const expectedData = [
      {id: 1, title: 'Post 1'},
      {id: 2, title: 'Post 2'},
      {id: 3, title: 'Post 3'},
    ];
    (useFetchPosts as jest.Mock).mockReturnValue({
      data: expectedData,
      loading: false,
      error: null,
    });
    const {getByText, toJSON} = render(<PostList userId="789" />);
    await act(async () => {
      expectedData.forEach(post => {
        expect(getByText(post.title)).toBeTruthy();
      });
    });
    expect(toJSON()).toMatchSnapshot();
  });
});
