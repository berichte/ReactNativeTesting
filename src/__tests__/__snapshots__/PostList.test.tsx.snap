// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList should render error state 1`] = `
<Text>
  Error: 
  Network error
</Text>
`;

exports[`PostList should render post list 1`] = `
<View>
  <View>
    <Text>
      Post 1
    </Text>
  </View>
  <View>
    <Text>
      Post 2
    </Text>
  </View>
  <View>
    <Text>
      Post 3
    </Text>
  </View>
</View>
`;
